lib/Sema/CSApply.cpp

```diff
@@ -2758,83 +2758,43 @@ namespace {
 
-        ConcreteDeclRef witness =
-            conformance.getWitnessByName(type->getRValueType(), constrName);
-        if (!witness || !isa<AbstractFunctionDecl>(witness.getDecl()))
-          return nullptr;
-        return witness;
-      };
-
-      auto *interpolationProto = TypeChecker::getProtocol(
-          cs.getASTContext(), expr->getLoc(),
-          KnownProtocolKind::ExpressibleByStringInterpolation);
-      auto associatedTypeDecl =
-          interpolationProto->getAssociatedType(ctx.Id_StringInterpolation);
-      if (associatedTypeDecl == nullptr) {
-        ctx.Diags.diagnose(expr->getStartLoc(),
-                           diag::interpolation_broken_proto);
+      ConcreteDeclRef witness =
+          conformance.getWitnessByName(type->getRValueType(), constrName);
+      if (!witness || !isa<AbstractFunctionDecl>(witness.getDecl()))
         return nullptr;
-      }
-      auto interpolationType =
-        simplifyType(DependentMemberType::get(openedType, associatedTypeDecl));
-
-      // Fetch needed witnesses.
-      ConcreteDeclRef builderInit = fetchProtocolInitWitness(
-          KnownProtocolKind::StringInterpolationProtocol, interpolationType,
-          {ctx.Id_literalCapacity, ctx.Id_interpolationCount});
-      if (!builderInit) return nullptr;
-      expr->setBuilderInit(builderInit);
-
-      ConcreteDeclRef resultInit = fetchProtocolInitWitness(
-          KnownProtocolKind::ExpressibleByStringInterpolation, type,
-          {ctx.Id_stringInterpolation});
-      if (!resultInit) return nullptr;
-      expr->setInitializer(resultInit);
-
-      // Make the integer literals for the parameters.
-      auto buildExprFromUnsigned = [&](unsigned value) {
-        LiteralExpr *expr = IntegerLiteralExpr::createFromUnsigned(ctx, value);
-        cs.setType(expr, ctx.getIntType());
-        return handleIntegerLiteralExpr(expr);
-      };
+      return witness;
+    }
 
-      expr->setLiteralCapacityExpr(
-          buildExprFromUnsigned(expr->getLiteralCapacity()));
-      expr->setInterpolationCountExpr(
-          buildExprFromUnsigned(expr->getInterpolationCount()));
+    Expr *
+    visitInterpolatedStringLiteralExpr(InterpolatedStringLiteralExpr *expr) {
+      auto &ctx = cs.getASTContext();
 
-      // This OpaqueValueExpr represents the result of builderInit above in
-      // silgen.
-      OpaqueValueExpr *interpolationExpr =
-          new (ctx) OpaqueValueExpr(expr->getSourceRange(), interpolationType);
-      cs.setType(interpolationExpr, interpolationType);
-      expr->setInterpolationExpr(interpolationExpr);
+      // Figure out the string type we're converting to.
+      auto ty = simplifyType(cs.getType(expr));
+      cs.setType(expr, ty);
 
-      auto appendingExpr = expr->getAppendingExpr();
-      appendingExpr->setSubExpr(interpolationExpr);
+      // Fetch the needed witness for the init(stringInterpolation:) call.
+      ConcreteDeclRef resultInit = fetchProtocolInitWitness(
```

Comment:
There is a (relatively) new constraint to do this during solving - `ValueWitness` so I think even this bit of code could be refactored?

Reply:
Ah nice, I'll take a look at that

Reply:
So it seems like this is something we may want to apply more generally to other literals, and perhaps even have it be a part of `LiteralConformsTo`, which I think would be best done as a follow-up